src/Tasks/ResolveAssemblyReferences/Server/RarController.cs

```diff
@@ -0,0 +1,158 @@
+
+        /// <summary>
+        /// Timeout for incoming connections
+        /// </summary>
+        private readonly TimeSpan Timeout = new TimeSpan(0, 15, 0);
+
+        public RarController(string pipeName, TimeSpan? timeout = null) : this(pipeName, timeout: timeout, resolveAssemblyReferenceTaskHandler: new RarTaskHandler())
+        {
+        }
+
+        internal RarController(string pipeName, IResolveAssemblyReferenceTaskHandler resolveAssemblyReferenceTaskHandler, TimeSpan? timeout = null)
+        {
+            _pipeName = pipeName;
+            _resolveAssemblyReferenceTaskHandler = resolveAssemblyReferenceTaskHandler;
+
+            if (timeout.HasValue)
+            {
+                Timeout = timeout.Value;
+            }
+        }
+
+        public async Task<int> StartAsync(CancellationToken cancellationToken = default)
+        {
+            using ServerMutex mutex = new ServerMutex(_pipeName);
+
+            if (!mutex.IsLocked)
+            {
+                return 1;
+            }
+
+            using CancellationTokenSource cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            CancellationToken token = cancellationTokenSource.Token;
+
+            while (!token.IsCancellationRequested)
+            {
+                // server will dispose stream too.
+                NamedPipeServerStream serverStream = GetStream(_pipeName);
+                await serverStream.WaitForConnectionAsync(token).ConfigureAwait(false);
+
+                // Connected! Refresh timeout for incoming request
+                cancellationTokenSource.CancelAfter(Timeout);
+
+                _ = HandleClientAsync(serverStream, token).ConfigureAwait(false);
+            }
+
+            return 0;
+        }
+
+        private async Task HandleClientAsync(Stream serverStream, CancellationToken cancellationToken = default)
+        {
+            using JsonRpc server = GetRpcServer(serverStream, _resolveAssemblyReferenceTaskHandler);
+            server.StartListening();
+
+            try
+            {
+                await server.Completion.WithCancellation(cancellationToken).ConfigureAwait(false);
+            }
+            catch (ConnectionLostException)
+            {
+                // Some problem with connection, let's ignore it.
+                // All other exceptions are issue though
+            }
+        }
+
+        private JsonRpc GetRpcServer(Stream stream, IResolveAssemblyReferenceTaskHandler handler)
+        {
+            IJsonRpcMessageHandler serverHandler = RpcUtils.GetRarMessageHandler(stream);
+            JsonRpc rpc = new JsonRpc(serverHandler, handler);
+            return rpc;
+        }
+
+        /// <summary>
+        /// Instantiates an endpoint to act as a client
+        /// </summary>
+        /// <param name="pipeName">The name of the pipe to which we should connect.</param>
+        private NamedPipeServerStream GetStream(string pipeName)
```

Comment:
Can this be implemented using `NamedPipeUtil.NamedPipeServerStream()`?

Reply:
I had thought about this too. But I would have to pass the `Func<>` in constructor to be able to use it, that is not problem, but I also need to create the `NamedPipeServerStream` with special right flag (`CreateNewInstance`). I could add parameter to that function to support that though... I will change it.

Reply:
Fixed